### PR TITLE
Fix missing MPI in custom-matrix-format example

### PR DESCRIPTION
**Pull request by MarcelKoch**
This PR fixes the missing MPI in the custom-matrix-format example when using the MPI wrapper as CXX compiler. There are actually two ways of fixing this implemented here:

1. remove the `<ginkgo/ginkgo.hpp>` include in the cuda file. Now the MPI headers are not required when compiling the cuda code anymore.
2. Use the CMake option `MPI_ASSUME_NO_BUILTIN_MPI to disable using MPI from the compiler wrapper directly.

Both of these fix the issue. I think having both of them is a good idea, because 1. the include is just not necessary, and 2. this would allow users to include `<ginkgo/ginkgo.hpp>` in device code.

### CODE REVIEW

**Comment by MarcelKoch** [comment]
Fixes part of #1367

**Comment by yhmtsai** [comment]
sorry, I do not get the usage of `MPI_ASSUME_NO_BUILTIN_MPI`.
could you elaborate it on two cases from gcc and mpicc?

File: cmake/autodetect_executors.cmake [hunk 1]
```diff
@@ -14,7 +14,10 @@ if (NOT DEFINED GINKGO_BUILD_OMP)
     endif()
 endif()
 
-if (NOT DEFINED GINKGO_BUILD_MPI)
+if(NOT DEFINED GINKGO_BUILD_MPI)
+    set(GINKGO_MPI_ASSUME_NO_BUILTIN_MPI ON CACHE BOOL "Disables using MPI wrapper compiler directly.")
+    mark_as_advanced(GIKNGO_MPI_ASSUME_NO_BUILTIN_MPI)
```
Comment (upsj) [inline cmake/autodetect_executors.cmake:19]: Do we need to make this controllable at all? It seems to be relevant only in very limited circumstances, where either `mpiexec` is missing from the environment, or the user has a MPI compiler wrapper loaded that is not the one specified by `CXX_COMPILER`

**Review by upsj — DISMISSED** [review]
LGTM! One minor nit (we already have way too many configuration options)

**Comment by MarcelKoch** [comment]
@yhmtsai If you set `CXX=mpicxx` then compiling with it doesn't require adding include paths or libraries for MPI. That gets picked up by cmake, when it finds the MPI package. So, on the cmake level, linking to MPI doesn't require any additional flags. This will lead to errors, when something is compiled not using `mpicxx`, but still including/linking to MPI. The `MPI_ASSUME_NO_BUILTIN_MPI` flag tells cmake to somehow ignore the MPI compiler wrapper and add the needed include path and libraries regardless.

File: cmake/autodetect_executors.cmake (hunk 1 above)
Comment (MarcelKoch) [reply cmake/autodetect_executors.cmake:19]: TBH, I'm not 100% sure how cmake finds MPI with this option on. Perhaps there are some weird cases, where it doesn't use the same MPI as the MPI wrapper would have. So I added the option to turn it off. I also marked it as advanced to indicate that users generally need not to worry about it.

**Comment by upsj** [comment]
This whole setting has me a bit confused - theoretically, CMake should interrogate the compiler wrapper for the location of the `mpi.h` file and add the include flag to the MPI target, but that doesn't seem to be happening here? From FindMPI.cmake: `# Given the new design of FindMPI, INCLUDE_DIRS will always be located, even under implicit linking.`

**Review by pratikvn — APPROVED** [review]

**Comment by MarcelKoch** [comment]
@upsj The comment from `FindMPI.cmake` is not in the official documentation, so I don't think we can rely on that. I tried again (without `MPI_ASSUME_NO_BUILTIN_MPI`) with the latest cmake release (3.26.4) and they still not set the include directories.
So maybe that is a bug in cmake, or the comment is just not very clear.

**Comment by upsj** [comment]
@MarcelKoch yes, I'd really like to get a full trace of the execution, because this seems like a bug to me. For this to be fixed, we need `MPI_CUDA_INCLUDE_DIRS` to be set, which seems to be missing.

**Comment by MarcelKoch** [comment]
@upsj I will open an cmake issue to see what's going on. From a brief look at the source, you're right, it always sets the include directories, but if a compiler wrapper is used, the value of the used variable is empty. So the source code comment is moot.

**Comment by MarcelKoch** [comment]
Here is the issue, let's see what happens: https://gitlab.kitware.com/cmake/cmake/-/issues/25085

**Comment by upsj** [comment]
I no longer think this fix is necessary, instead we should be propagating the C++ compiler as CUDA host compiler correctly.

**Comment by MarcelKoch** [comment]
Fixed by #1368 and #1205

**Pull request closed without merging**